plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -0,0 +1,618 @@
+		return FALSE;
+	if (!fu_rts54hub_rtd21xx_device_read_status_raw (self, &status, error))
+		return FALSE;
+	if (status != ISP_STATUS_IDLE_SUCCESS) {
+		g_set_error (error,
+				FWUPD_ERROR,
+				FWUPD_ERROR_INTERNAL,
+				"detach status was 0x%02x", status);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_setup (FuDevice *device, GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* get version */
+	locker = fu_device_locker_new_full (device,
+					    (FuDeviceLockerFunc) fu_device_detach,
+					    (FuDeviceLockerFunc) fu_device_attach,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_rts54hub_rtd21xx_ensure_version_unlocked (self, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_reload (FuDevice *device, GError **error)
+{
+	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE (fu_device_get_parent (device));
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open parent device */
+	locker = fu_device_locker_new (parent, error);
+	if (locker == NULL)
+		return FALSE;
+	return fu_rts54hub_rtd21xx_device_setup (device, error);
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_detach (FuDevice *device, GError **error)
+{
+	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE (fu_device_get_parent (device));
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open device */
+	locker = fu_device_locker_new (parent, error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_device_retry (device,
+			      fu_rts54hub_rtd21xx_device_detach_cb,
+			      100, NULL, error))
+		return FALSE;
+
+	/* success */
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_attach (FuDevice *device, GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE (fu_device_get_parent (device));
+	guint8 buf[] = { ISP_CMD_FW_UPDATE_RESET };
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open device */
+	locker = fu_device_locker_new (parent, error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_FOREGROUND_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   buf, sizeof(buf),
+						   error)) {
+		g_prefix_error (error, "failed to attach: ");
+		return FALSE;
+	}
+
+	/* success */
+	fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_write_firmware (FuDevice *device,
+					   FuFirmware *firmware,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	const guint8 *fwbuf;
+	gsize fwbufsz = 0;
+	guint32 project_addr;
+	guint8 project_id_count;
+	guint8 read_buf[10] = { 0 };
```

Comment:
inconsistent use of spacing between this and next variable

Reply:
Fixed.